File: ugly2.py
"""
Time: O(n)
Space:O(n)
Leet: Accepted
Problems: Could not come up with Solution had to refer leetcode Solution
"""
class Solution(object):
    def nthUglyNumber(self, n):
        """
        :type n: int
        :rtype: int
        """
        ugly = [1] #initiate array
        ptr2=ptr3=ptr5=0 #maintain 3 pointers

        for i in range(1,n):
            u2,u3,u5 = 2*ugly[ptr2], 3*ugly[ptr3], 5*ugly[ptr5] #derive the 3 next possible values
            nex = min(u2,u3,u5)
            ugly.append(nex) #append the correct one

            #increment whichever pointer is appended
            if u2 == nex:
                ptr2 += 1
            if u3 == nex:
                ptr3 += 1
            if u5 == nex:
                ptr5 += 1

        return ugly[-1]

File: Ugly2.java
/*
Time Complexity - O(n) where is the input number n.
Space Complexity - O(n)
*/
class Solution {
    public int nthUglyNumber(int n) {
        int[] ugly = new int[n];
        int nextMul2 = 2, nextMul3 = 3, nextMul5 = 5;
        int index2 = 1, index3 = 1, index5 = 1;

        ugly[0] = 1;
        for(int i = 1; i < n; i++) {
            ugly[i] = Math.min(nextMul2, Math.min(nextMul3, nextMul5));
            if(ugly[i] == nextMul2) 
                nextMul2 = ugly[index2++] * 2;
            if(ugly[i] == nextMul3)
                nextMul3 = ugly[index3++] * 3;
            if(ugly[i] == nextMul5)
                nextMul5 = ugly[index5++] * 5;
        }
        return ugly[n - 1];
    }
}
